processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -52,6 +51,11 @@ public OnheapIncrementalIndex(IncrementalIndexSchema incrementalIndexSchema, boo
     super(incrementalIndexSchema, deserializeComplexMetrics);
     this.facts = new ConcurrentSkipListMap<>();
     this.maxRowCount = maxRowCount;
+
+    this.aggList = new Aggregator[maxRowCount][];// To handle concurrency, we allocate up to maxRowCount
```

Comment:
I think this is overkill with memory allocation. We can revert back to the model we had in 0.6 and used a thread safe data structure here.